sys/kern/uipc_shm.c

```diff
@@ -1046,6 +1047,104 @@ shm_vmspace_free(struct vmspace *vm)
 	vm_map_unlock(&vm->vm_map);
 }
 
+void
+shm_map_local_objs(struct proc *p, struct vm_cheri_revoke_cookie *crc)
+{
+	struct vmspace *vm = p->p_vmspace;
+	vm_map_t map = &vm->vm_map;
+	struct shmfd *shmfd;
+	void *rl_cookie;
+	vm_prot_t prot, max_prot;
+	int rv;
+
+	vm_map_lock(map);
+	LIST_FOREACH(shmfd, &vm->vm_shm_objects, shm_vmspace_entry) {
+		vm_offset_t align = PAGE_SIZE;
+		vm_offset_t vaddr = 0;
+		vm_pointer_t reservation;
+		/* XXX: can we avoid mapping some objects? */
+
+		prot = max_prot = VM_PROT_ALL;
+
+		shm_hold(shmfd);
+		rl_cookie = shm_rangelock_wlock(shmfd, 0, OFF_MAX);
+
+		vm_object_reference(shmfd->shm_object);
+
+		KASSERT(shmfd->shm_hoard_addr == 0,
+		    ("shm obj already mapped?"));
+		if (shm_largepage(shmfd))
+			align = pagesizes[shmfd->shm_lp_psind];
+		align = MAX(align,
+		    CHERI_REPRESENTABLE_ALIGNMENT(shmfd->shm_size));
+
+		/*
+		 * We don't need to do the retry dance to try and avoid sbrk
+		 * space because this only applies to pure-capability
+		 * processes and those don't do sbrk.
+		 */
+		rv = vm_map_find_aligned(map, &vaddr,
+		    shmfd->shm_size, vm_map_max(map), align);
+		if (rv != KERN_SUCCESS)
+			goto fail;
+		reservation = vaddr;
+		rv = vm_map_reservation_create_locked(map,
+		    &reservation, shmfd->shm_size, max_prot);
+		if (rv != KERN_SUCCESS)
+			goto fail;
+		rv = vm_map_insert(map, shmfd->shm_object, 0,
```

Comment:
Is there a particular reason we have to map them in order to scan them, or is this just more convenient? In general I'd expect it to be significantly cheaper to scan the resident pages directly (and page in as needed, based on the capability tags stashed in swap block structures). That approach however does require more code, so I understand why you went with this approach.

Reply:
The existing scan logic is more `vm_map_entry`-centric than might be ideal.  Perhaps `‎vm_cheri_revoke_object_at` should be refactored to not take the `entry` and instead take (and return) the fields it needs.  Then we could sweep an arbitrary chunk of an object, mapped or not.

Though that approach is at odds with an experiment I'd love to have time to run: https://github.com/CTSRD-CHERI/paper-revocation3/issues/33

Reply:
I did it because it was easy and took advantage of existing infrastructure. Ideally we'd probably scan the pages directly at the end and thus do minimal work in the common case where the pages were already mapped.